### ISSUE TITLE
Unsafe concurrent "SingularTable" method call

### ISSUE THREAD

**Issue by bazuka** [jinzhu/gorm#1526]
### What version of Go are you using (`go version`)?
go version go1.8.1 linux/amd64

### Which database and its version are you using?
postgresql 9.6.1

### What did you do?
When using the value of GOMAXPROCS > 1, we noticed error messages "can't preload field `Field` for `Model`".
After investigation i've found that sometimes `Field.Relationship == nil` which led to these errors . 
I continued to dig deeper and found the cause of the error.
In our database we are using singular table names. 
So for each request we are open new connection with method `Init` 
```go
func (manager *DBManager) Init()(*gorm.DB, error){
	gdb, err := gorm.Open("postgres", manager.getConnectionString())
	gdb.SingularTable(true)
        return gdb, err
}
```
And under the hood method `db.SingularTable(true)` is resetting the `modelStructureMap` variable  which is global.
```go
//main.go#L159
func (s *DB) SingularTable(enable bool) {
   modelStructsMap = newModelStructsMap()
   s.parent.singularTable = enable
}
```
So, in case of multiple concurrent requests we have corrupted `modelStructureMap`.

Possible desicion is remove cache variable invalidation in `db.SingularTable(true)` since SingularTable flag is should be set once in connection initialization step and cache already initialized.

I've tried to comment this [line](https://github.com/jinzhu/gorm/blob/master/main.go#L160) locally and it resolves the problem.

Another option is, as it was implemented first time(commit  73a0401), cache could be scoped to current DB connection.

**Comment by kpeu3i** [comment]
Moreover, if I turn on SingularTable mode, I get an error `can't preload field Field for Model"` even `GOMAXPROCS=1`

**Comment by kpeu3i** [comment]
Ping!

Houston we have a problem :)

**Comment by emirb** [comment]
Fixed in #2396 and released in [v1.9.4](https://github.com/jinzhu/gorm/releases/tag/v1.9.4).
